pkg/cloudprovider/vmss.go

```diff
@@ -121,41 +117,80 @@ func (h *vmssIdentityHolder) IdentityInfo() IdentityInfo {
 }
 
 func (h *vmssIdentityHolder) ResetIdentity() IdentityInfo {
-	h.vmss.Identity = &compute.VirtualMachineScaleSetIdentity{}
+	h.vmss.Identity = &compute.VirtualMachineScaleSetIdentity{
+		Type:                   compute.ResourceIdentityTypeUserAssigned,
+		UserAssignedIdentities: make(map[string]*compute.VirtualMachineScaleSetIdentityUserAssignedIdentitiesValue),
+	}
 	return h.IdentityInfo()
 }
 
 type vmssIdentityInfo struct {
 	info *compute.VirtualMachineScaleSetIdentity
 }
 
-func (i *vmssIdentityInfo) RemoveUserIdentity(id string) error {
-	if err := filterUserIdentity(&i.info.Type, i.info.IdentityIds, id); err != nil {
-		return err
-	}
-	// If we have either no identity assigned or have the system assigned identity only, then we need to set the
-	// IdentityIds list as nil.
-	if i.info.Type == compute.ResourceIdentityTypeNone || i.info.Type == compute.ResourceIdentityTypeSystemAssigned {
-		i.info.IdentityIds = nil
+func (i *vmssIdentityInfo) GetUserIdentityList() []string {
+	var ids []string
+	if i.info == nil {
+		return ids
 	}
-	// if the identityids is nil and identity type is not set, then set it to ResourceIdentityTypeNone
-	if i.info.IdentityIds == nil && i.info.Type == "" {
-		i.info.Type = compute.ResourceIdentityTypeNone
+	for id := range i.info.UserAssignedIdentities {
+		ids = append(ids, id)
 	}
-	return nil
+	return ids
 }
 
-func (i *vmssIdentityInfo) AppendUserIdentity(id string) bool {
-	if i.info.IdentityIds == nil {
-		var ids []string
-		i.info.IdentityIds = &ids
+func (i *vmssIdentityInfo) SetUserIdentities(ids map[string]bool) bool {
+	expectedIDList := make(map[string]bool)
+
+	if i.info.UserAssignedIdentities == nil {
+		i.info.UserAssignedIdentities = make(map[string]*compute.VirtualMachineScaleSetIdentityUserAssignedIdentitiesValue)
 	}
-	return appendUserIdentity(&i.info.Type, i.info.IdentityIds, id)
-}
 
-func (i *vmssIdentityInfo) GetUserIdentityList() []string {
-	if i.info.IdentityIds == nil {
-		return []string{}
+	for id, _ := range i.info.UserAssignedIdentities {
+		expectedIDList[id] = true
+	}
+
+	for id, add := range ids {
+		_, exists := expectedIDList[id]
+
+		// identity already exists/not exists, so skip doing add/del operation for
+		// the identity again for the node
+		if (exists && add) || (!exists && !add) {
+			delete(ids, id)
+		}
+		// already exists on node and want to remove existing identity
```

Comment:
Could we simplify the logic here to use either expectedIDList or ids. Is it possible to combine the logic here and the last loop and avoid to iterate again and do modifications to the ids ? And then return in case of empty node ids ?